src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.cs

```diff
@@ -2,47 +2,239 @@
-                return false;
+                if (maxThreads < s_minThreads || maxThreads == 0)
+                {
+                    return false;
+                }
+                else
+                {
+                    short threads = (short)Math.Min(maxThreads, MaxPossibleThreadCount);
+                    if (s_forcedMaxWorkerThreads == 0)
+                    {
+                        s_maxThreads = threads;
+
+                        ThreadCounts counts = ThreadCounts.VolatileReadCounts(ref s_separated.counts);
+                        while (counts.numThreadsGoal > s_maxThreads)
+                        {
+                            ThreadCounts newCounts = counts;
+                            newCounts.numThreadsGoal = s_maxThreads;
+
+                            ThreadCounts oldCounts = ThreadCounts.CompareExchangeCounts(ref s_separated.counts, newCounts, counts);
+                            if (oldCounts == counts)
+                            {
+                                counts = newCounts;
+                            }
+                            else
+                            {
+                                counts = oldCounts;
+                            }
+                        }
+                    }
+                    return true;
+                }
             }
-            else
+            finally
             {
-                s_maxThreads = threads;
-                return true;
+                s_maxMinThreadLock.Release();
             }
         }
 
         public static int GetMaxThreads() => s_maxThreads;
+
+        public static int GetAvailableThreads()
+        {
+            ThreadCounts counts = ThreadCounts.VolatileReadCounts(ref s_separated.counts);
+            return counts.numExistingThreads - counts.numProcessingWork;
+        }
+
+        internal static bool NotifyWorkItemComplete()
+        {
+            // TODO: Check perf. Might need to make this thread-local.
+            Interlocked.Increment(ref s_completionCount);
+            Volatile.Write(ref s_separated.lastDequeueTime, Environment.TickCount);
+
+            bool shouldAdjustWorkers = ShouldAdjustMaxWorkersActive();
+            if(shouldAdjustWorkers)
+            {
+                AdjustMaxWorkersActive();
+            }
+            return !WorkerThread.ShouldStopProcessingWorkNow();
+        }
+
+        //
+        // This method must only be called if ShouldAdjustMaxWorkersActive has returned true, *and*
+        // s_threadAdjustmentLock is held.
+        //
+        private static void AdjustMaxWorkersActive()
+        {
+            int currentTicks = Environment.TickCount;
+            int totalNumCompletions = Volatile.Read(ref s_completionCount);
+            int numCompletions = totalNumCompletions - s_separated.priorCompletionCount;
+            long startTime = s_currentSampleStartTime;
+            long endTime = 0; // TODO: PAL High Performance Counter
+            long freq = 0;
```

Comment:
You may need to initialize these to some values that work if you're going to hook up ClrThreadPool to the actual thread pool, otherwise it would divide by zero below